File: POTD/Graphs/Union Find/Python Sols/1722. Minimize Hamming Distance After Swap Operations.py

class disjointSet:
    def __init__(self,n:int):
         self.parent=list(range(n))
         self.rank=[0] * n
    
    def findUPar(self,node:int)->int:
        if node==self.parent[node]:
            return node
        
        self.parent[node]=self.findUPar(self.parent[node])
        return self.parent[node]
    
    
    def unionByRank(self,u:int,v:int):
        pu=self.findUPar(u)
        pv=self.findUPar(v)

        if pu==pv:
            return
        
        if self.rank[pu] < self.rank[pv]:
            self.parent[pu]=pv
        elif self.rank[pv] < self.rank[pu]:
            self.parent[pv]=pu
        else:
            self.parent[pu]=pv
            self.rank[pv]+=1
    

        

class Solution:
    def minimumHammingDistance(self, source: List[int], target: List[int], allowedSwaps: List[List[int]]) -> int:
        size=len(source)

        ds = disjointSet(size)

        for u,v in allowedSwaps:
            ds.unionByRank(u,v)
        
        mp= collections.defaultdict(list)
        for i in range(size):
            mp[ds.findUPar(i)].append(i)
        
        minHamDist=0

        for p,vec in mp.items():
            reducer=collections.defaultdict(int)

            for idx in vec:
                reducer[source[idx]]+=1
            
            # now reduce

            for idx in vec:
                if reducer[target[idx]]>0:
                    reducer[target[idx]]-=1
                else:
                    minHamDist+=1
            
        
        return minHamDist

File: POTD/Graphs/Union Find/Cpp Sols/1722. Minimize Hamming Distance After Swap Operations.cpp
// TIME : O((N+M)*ALPHA(N))
// SPACE : O(N)

class disjointSet{
    vector<int> parent;
    vector<int> rank;

    public:
    disjointSet(int n){
        parent.resize(n);
        rank.resize(n,0);

        for(int i=0;i<n;i++){
            parent[i]=i;
        }
    }


    int findUPar(int node){
        if(node==parent[node]) return node;

        return parent[node]=findUPar(parent[node]);
    }


    void unionByRank(int u,int v){
        int pu=findUPar(u);
        int pv=findUPar(v);

        if(pu==pv) return;

        if(rank[pu]<rank[pv]){
            parent[pu]=pv;
        }
        else if(rank[pv]<rank[pu]){
            parent[pv]=pu;
        }
        else{
            parent[pu]=pv;
            rank[pv]+=1;
        }
    }
};

class Solution {
public:
    int minimumHammingDistance(vector<int>& source, vector<int>& target, vector<vector<int>>& allowedSwaps) {
        int size=source.size();
        disjointSet ds(size);

        for(auto &swap:allowedSwaps){
            ds.unionByRank(swap[0],swap[1]);
        }

        unordered_map<int,vector<int>> mp;

        for(int i=0;i<size;i++){
            mp[ds.findUPar(i)].push_back(i);
        }

        int minHamDist=0;

        for(auto &it:mp){
            auto &vec=it.second;

            unordered_map<int,int> reducer;

            for(auto &idx:vec){
                // first populate from source
                reducer[source[idx]]++;
            }

            // then reduce from target
            for(auto&idx:vec){
                if(reducer[target[idx]]>0){
                    reducer[target[idx]]--;
                }
                else{
                    minHamDist++;
                }
            }
        }

        return minHamDist;
    }
};
